File: chapter4/hanoi_ricorsivo/Java/HanoiRicorsivo.java
/*
 * Chapter 4
 * Paragraph 4
 *
 * Hanoi-ricorsiva
 */

public class HanoiRicorsivo {
    public static void doHanoi(int n, int origine, int destinazione, int intermedio) {
        if (n == 1) {
            System.out.println("Trasferisci un disco da " + origine + " a " + destinazione);
        }
        else {
            doHanoi(n-1, origine, intermedio, destinazione);
            System.out.println("Trasferisci un disco da " + origine + " a " + destinazione);
            doHanoi(n-1, intermedio, destinazione, origine);
        }
    }

    public static void main(String[] args) {
        doHanoi(3, 1, 3, 2);
    }
}

File: chapter04/hanoi_ricorsivo/C++/hanoi_ricorsivo.cpp
/*
 * Chapter 4
 * Paragraph 4
 *
 * Hanoi-ricorsiva
 */

#include <iostream>
using namespace std;

void hanoi_ricorsiva(int n, int origine, int destinazione, int intermedio) {
  if (n == 1) {
    cout << "Trasferisci un disco da " << origine << " a " << destinazione << endl;
  }
  else {
    hanoi_ricorsiva(n-1, origine, intermedio, destinazione);
    cout << "Trasferisci un disco da " << origine << " a " << destinazione << endl;
    hanoi_ricorsiva(n-1, intermedio, destinazione, origine);
  }
}

int main()
{
  hanoi_ricorsiva(3, 1, 3, 2);
  return 0;
}
